src/psset.c

```diff
@@ -195,6 +196,51 @@ psset_alloc_container_remove(psset_t *psset, hpdata_t *ps) {
 	}
 }
 
+static size_t
+psset_purge_list_ind(hpdata_t *ps) {
+	size_t ndirty = hpdata_ndirty_get(ps);
+	/* Shouldn't have something with no dirty pages purgeable. */
+	assert(ndirty > 0);
+	pszind_t pind = sz_psz2ind(sz_psz_quantize_floor(ndirty << LG_PAGE));
+	/*
+	 * Higher indices correspond to lists we'd like to purge earlier;
+	 * increment the index for the nonhugified hpdatas first, so that we'll
+	 * pick them before picking hugified ones.
+	 */
+	return (size_t)pind * 2 + (hpdata_huge_get(ps) ? 0 : 1);
+}
+
+static void
+psset_maybe_remove_purge_list(psset_t *psset, hpdata_t *ps) {
+	/*
+	 * Remove the hpdata from its purge list (if it's in one).  Even if it's
+	 * going to stay in the same one, by appending it during
+	 * psset_update_end, we move it to the end of its queue, so that we
+	 * purge LRU within a given dirtiness bucket.
+	 */
+	if (hpdata_purge_allowed_get(ps)) {
```

Comment:
Question: in what case do we go down this path while purge is not allowed? (same question for the insert below). Does that mean setting purge allowed and the list operation need to be ordered?

Reply:
I'm not totally sure I follow; a pageslab is never in the purge list in between psset_update_begin and psset_update_end. So in that sense, the operations need to be ordered, but not any more so than how every pageslab metadata update happens between begin and end.

Reply:
If we want to disallow a pageslab (which is currently in the purge list) from being purged, the order should be, remove it from the purge list first, and then update the metadata for `purge_allowed`? Because otherwise if we update the metadata first, this function would simply be skipped entirely?

Reply:
> a pageslab is never in the purge list in between psset_update_begin and psset_update_end

Is this ensured by the API or the logic flow like we always take a pageslab off the lists before, or it's disallowed since it doesn't make sense? I might be missing something here; is it right that a pageslab on the purge list can be partially in use, and keeps serving requests, and be a part of the psset that way? (which is different than PAC "purge list").

Reply:
Discussed offline; I forgot that the LRU-nature of the purgelist means that every time a pageslab gets updated it will also needs to be adjusted for the LRU.

Do we have assertions to check for that case, i.e. pageslab not on purge list during update (I seem to recall we do)?

Reply:
Not for purging (I don't think there's in general an easy way to check if a pageslab is in a list in our linked list implementation). So the "can purge" bool tells us if something is in the list when it's not being updated.

Reply:
Sounds good. Thanks!